### DISCUSSION TITLE
Accessing blockchain data from state machine

### DISCUSSION THREAD

**Discussion by robert-zaremba** [tendermint/tendermint#5977] [Ideas]
Moving from: https://github.com/cosmos/cosmos-sdk/discussions/8272

## Rationale

Currently, State Machine Programs (SMPs) [*] can't query blockchain data (transactions, block headers) from past blocks.
There are several use-cases for that:
* a SMP would like to see block voting stats
* a SMP would like to examine transaction data
* some validator operations
* timestamping data: we make a transaction, and use _calldata_ (message attributes which are not processed otherwise) - there is no need to copy that to a state, because messages are already stored in the block.

The workaround is to copy transaction data into smart-contract state... which is not ideal, because we essentially duplicate a data.

#### State Machine Program (SMP)

To avoid any confusion, State Machine Program is an on chain app which encapsulates on-chain state logic. This could be: SDK module, comswasm smart contract, agoric smart contract, state-machine module....

## Goals

+ Estimate the efforts to allow app accessing blockchain data
+ See if there are any objections for that.

## References:

+ https://github.com/tendermint/spec/issues/203

**Comment by robert-zaremba** [comment]
@alessio , @tessr  - would it be possible to schedule it for a next release?

**Comment by tac0turtle** [comment]
Definitely think this is very useful. Currently, the way to do this is to query any data through the RPC but this seems like a simple workaround. Building this into the abci layer could be very useful. 

Note: To help avoid confusion, in Tendermint nomenclature this `Estimate the efforts to allow smart contracts accessing blockchain data` would translate to the application querying the Tendermint block store and state.

**Reply by alexanderbez** [reply]
Is this not what `ABCI#Query` is for? Why can't modules just use that?

**Reply by robert-zaremba** [reply]
`abci.Query` is for state query, not a blockchain data (eg search for transaction by tx hash, or show the block header by block height).

**Reply by alexanderbez** [reply]
How do you plan on exposing a reliable interface that serves state and metadata?

**Reply by robert-zaremba** [reply]
This is the goal of this discussion to figure it out. I'm not familiar with tendermint code nor API.

**Reply by alexanderbez** [reply]
Well the issue becomes these calls cannot be networked if the state machine depends on them for execution of any kind.

**Reply by tac0turtle** [reply]
`abci.Query` is for Tendermint to make requests to the application, not inverse. There wouldn't be a way for the application to call Tendermint. To do something like this we would need to provide a secondary flow where the application could make requests to tendermint. 

I do agree that the state machine using it for internal purposes makes this complicated. With pruning this blows up in complication. It can be done, but we would need to be extra careful.

**Reply by alexanderbez** [reply]
Well, it kind of is. It's app -> tendermint -> app -> response. Technically, you could call tendermint directly. But regardless, I just don't see how this can be done especially since you can run Tendermint out of process.

**Reply by tac0turtle** [reply]
> It's app -> tendermint -> app -> response.

This is app requests through RPC then tendermint makes a request to the app for some info and gets a response? This doesn't get info like a past header into the application, unless I missed something. 

We may need to introduce a secondary connection to allow apps to query Tendermint state, otherwise we tell apps to query the RPC directly as a workaround. (Not preferred as teams have mentioned in the past)

**Comment by michaelfig** [comment]
@marbar3778 I have a much more limited problem: how to provide synchronous access to ABCI events that were generated by other Cosmos modules within the same block?  Notably, this is information that has not yet been committed (thus it is not available via RPC).

This is not smart-contract specific, it is an alternative to having to modify a Cosmos module to expose particular hooks when its useful transitions are already exposed as events.

My experiment cosmos/cosmos-sdk#9305 solves this by keeping a copy of the current events within the Cosmos Context.

Should I take this back to cosmos/cosmos-sdk#8272 ?  Or should it usefully be part of Tendermint/ABCI?

**Reply by alexanderbez** [reply]
You should certainly do this in the SDK. Is there a reason you cannot use the `EventManager` on the `Context`?

**Reply by michaelfig** [reply]
The `EventManager`s events get cleared after each tx and so nothing is available for `EndBlocker`.  Hence cosmos/cosmos-sdk#9305 PR to patch the `EventManager` to keep a copy of them during each block.

Maybe we should take this thread back to cosmos-sdk.

**Reply by tac0turtle** [reply]
Robert uses the wording smart contract here, but he means state machine. 

We can move the conversation or open a new discussion in the sdk, if the current one does not fit your needs. 

The events are not within consensus, relying on them in the application can lead to non deterministic outputs.

**Reply by robert-zaremba** [reply]
We should use events in state machine / smart contracts (yes, in the description I'm referring to state machine, will update). They are not part of the consensus and can be wiped off.
So I agree with @alexanderbez  - if you need to do something with events in modules then it should not go beyond SDK boundary.

**Reply by robert-zaremba** [reply]
ps: oh, I updated it long time ago :sweat_smile:

**Reply by alexanderbez** [reply]
Ahhh I see @michaelfig, thanks for clarifying. Yeah, I think your idea is reasonable -- keep the `EventManager` logic as-is, but also allow for a non-purged version to exist up and until `EndBlock` that you can query and it'll be cleared after `EndBlock`.
